src/NLog.Loki/HttpLokiTransport.cs

```diff
@@ -49,19 +49,25 @@ public async Task WriteLogEventsAsync(LokiEvent lokiEvent)
     /// Prepares the HttpContent for the loki event(s).
     /// If gzip compression is enabled, prepares a gzip stream with the appropriate headers.
     /// </summary>
-    private HttpContent CreateContent<T>(T lokiEvent)
+    private async Task<HttpContent> CreateContent<T>(T lokiEvent)
     {
-        var jsonContent = JsonContent.Create(lokiEvent, options: _jsonOptions);
+        //var jsonContent = JsonContent.Create(lokiEvent, options: _jsonOptions);
+
+        using var memStream = new MemoryStream();
```

Comment:
Use a pooled memory stream to limit allocations.
https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream

Though, I wonder if we really need to preallocate the json before sending the HTTP request. Is there a way to stream the json serialization and compression without having to pre-allocate the json?

Reply:
It doesn't look like it's possible to write to the stream without pre-allocating the entirety of the json stream.
https://marcroussy.com/2020/11/02/serialization-with-system-text-json/

The pooled memory stream should still come handy.